### ISSUE TITLE
Incorrect report of undeclared label

### ISSUE THREAD

**Issue by s0lst1ce** [rust-lang/rust-analyzer#15516]
**rust-analyzer version**: 0.3.1631-standalone

**rustc version**: rustc 1.72.0 (5680fa18f 2023-08-23)

**relevant settings**: I don't believe I've set anything particular. I can share the rust-analyzer settings file if need be.
Hello, I'm having an issue with ra where it reports a ```use of undeclared label `'my_label` ``` (which I easily visualize with error lens on code).
Neither `cargo check` nor `clippy` report any warning (much less error as is the case with ra) regarding this when I run them in my terminal.
I'm not sure about a minimal example but here's one I quickly crafted from the actual snippet that causing me the issue.
```rs
struct PlaceHolder;

impl Iterator for PlaceHolder {
    type Item = Result<Option<()>, ()>;

    fn next(&mut self) -> Option<Self::Item> {
        Some(Ok(Some(())))
    }
}

fn main() {
    let mut placeholder = PlaceHolder;
    'my_label: while let Some(a) = match placeholder.next().unwrap() {
        Ok(_) => todo!(),
        Err(err) => {
            continue 'my_label; // use of undeclared label `'my_label`
        }
    } {
        a
    }
}
```

![image](https://github.com/rust-lang/rust-analyzer/assets/46009999/e0b90d27-976c-4143-bfc9-409f85c4d481)

**Comment by s0lst1ce** [comment]
And though this example does come with its share of warnings `cargo clippy | grep error | wc -l` outputs `0`.

**Comment by xffxff** [comment]
It looks like RA is unable to resolve the label in the match arm

```rust
fn main() {
    let mut optional = Some(0);

    'my_label: while let Some(a) = match optional {
        None => break 'my_label,
        Some(val) => Some(val),
    } {
        println!("{}", a);
        optional = None;
        continue 'my_label;
    }
}
```
![image](https://github.com/rust-lang/rust-analyzer/assets/30254428/c3899cb0-88f7-4583-a219-992887c1db54)
